idaes/models/unit_models/mscontactor.py

```diff
@@ -1246,11 +1394,78 @@ def _pressure_balance_rule(blk, t, s, stream, uom):
     if pconfig.has_pressure_change:
         rhs += getattr(blk, stream + "_deltaP")[t, s]
 
-    return 0 == rhs
+    return 0 * uom.PRESSURE == rhs
 
 
 def _side_stream_pressure_rule(b, t, s, stream):
     stage_state = getattr(b, stream)[t, s]
     side_state = getattr(b, stream + "_side_stream_state")[t, s]
 
     return stage_state.pressure == side_state.pressure
+
+
+def _holdup_rule(b, t, e, p, j, stream):
+    holdup = getattr(b, stream + "_material_holdup")
+    stage_state = getattr(b, stream)[t, e]
+    phase_frac = getattr(b, stream + "_phase_fraction")
+
+    return holdup[t, e, p, j] == (
+        b.volume[e]
+        * b.volume_frac_stream[t, e, stream]
+        * phase_frac[t, e, p]
+        * stage_state.get_material_density_terms(p, j)
+    )
+
+
+def _energy_holdup_rule(b, t, e, p, stream):
+    holdup = getattr(b, stream + "_energy_holdup")
+    stage_state = getattr(b, stream)[t, e]
+    phase_frac = getattr(b, stream + "_phase_fraction")
+
+    return holdup[t, e, p] == (
+        b.volume[e]
+        * b.volume_frac_stream[t, e, stream]
+        * phase_frac[t, e, p]
+        * stage_state.get_energy_density_terms(p)
+    )
+
+
+def _add_phase_fractions(b, stream, phase_list):
+    if len(phase_list) > 1:
+        phase_fraction = Var(
+            b.flowsheet().time,
+            b.elements,
+            phase_list,
+            initialize=1 / len(phase_list),
+            doc=f"Volume fraction of holdup by phase in stream {stream}",
+        )
+        b.add_component(stream + "_phase_fraction", phase_fraction)
+
+        sum_of_phase_fractions = Constraint(
+            b.flowsheet().time,
+            b.elements,
+            rule=partial(
+                _sum_phase_frac_rule, phase_frac=phase_fraction, phase_list=phase_list
+            ),
+            doc=f"Sum of phase fractions constraint for stream {stream}",
+        )
+        b.add_component(stream + "_sum_phase_fractions", sum_of_phase_fractions)
+
```

Comment:
Phase volume fraction needs to be given by the property package, and appropriate constraints need to be written here. 

```
def phase_volume_frac_rule(b, p):
    return b.flow_mol_phase[p] * b.dens_mol_phase[p] / sum( b.flow_mol_phase[q] * b.dens_mol_phase[q] for q in b.params.phase_set)
```

Reply:
Actually, I got mixed up, it should be 
````
def phase_volume_frac_rule(b, p):
    return b.flow_mol_phase[p] * b.vol_mol_phase[p] / sum( b.flow_mol_phase[q] * b.vol_mol_phase[q] for q in b.params.phase_set)
```